### PR TITLE
Added support for Array.contains and List.contains

### PR DESCRIPTION
**Pull request by faldor20**
I have completed the partially complete support for the "IN" keyword.
Array.contains and List.contains can both compare a list of local values against a single value from the table
They generate query code like the following:
```
a.tableVal IN("val1","val2")
```

I also added support for those same Array.contains and List.contains operators to work against table values
This would generate code like:
```
contains(a.tableList,"b")
```

### CODE REVIEW

File: src/FSharp.AWS.DynamoDB/Expression/ConditionalExpr.fs [hunk 1]
```diff
@@ -192,36 +192,48 @@ let extractQueryExpr (recordInfo : RecordTableInfo) (expr : Expr) : ConditionalE
...
-            | SpecificCall2 <@ Array.length @> (None, _, _, [AttributeGet attr]) -> 
+            | SpecificCall2 <@ Array.length @> (None, _, _, [AttributeGet attr]) ->
                 SizeOf attr.Id
 
             | _ -> invalidQuery()
 
+        ///Attempts to convert an expression into a list of operands.
+        ///This should be used when we know the expression is some kind of sequence.
+        let (|SequenceGet|_|) (expr:Expr)=
+            let op=extractOperand None expr
```
Comment (bartelink) [inline src/FSharp.AWS.DynamoDB/Expression/ConditionalExpr.fs:228]:
```suggestion
            let op = extractOperand None expr
```

File: src/FSharp.AWS.DynamoDB/Expression/ConditionalExpr.fs [hunk 2]
```diff
@@ -192,36 +192,48 @@ let extractQueryExpr (recordInfo : RecordTableInfo) (expr : Expr) : ConditionalE
...
+            match op with
+            |Value va ->
+                va.AttributeValue.L
+                |>Seq.map(fun x->x|>wrap|>Value)
```
Comment (bartelink) [inline src/FSharp.AWS.DynamoDB/Expression/ConditionalExpr.fs:232]: Hey, we can afford a few more bits!
```suggestion
                |> Seq.map (fun x-> x |> wrap |> Value)
```

File: tests/FSharp.AWS.DynamoDB.Tests/ConditionalExpressionTests.fs [hunk 3]
```diff
@@ -82,6 +83,7 @@ type ``Conditional Expression Tests`` (fixture : TableFixture) =
             Map = seq { for _ in 0L .. rand() % 5L -> "K" + guid(), rand() } |> Map.ofSeq
             Set = seq { for _ in 0L .. rand() % 5L -> rand() } |> Set.ofSeq
             List = [for _ in 0L .. rand() % 5L -> rand() ]
+            Array =  [|for _ in 0L .. rand() % 5L -> rand() |]
```
Comment (bartelink) [inline tests/FSharp.AWS.DynamoDB.Tests/ConditionalExpressionTests.fs:86]:
```suggestion
            Array = [| for _ in 0L .. rand() % 5L -> rand() |]
```

File: tests/FSharp.AWS.DynamoDB.Tests/ConditionalExpressionTests.fs [hunk 4]
```diff
@@ -330,6 +332,7 @@ type ``Conditional Expression Tests`` (fixture : TableFixture) =
         table.PutItem(item, <@ fun r -> r.Set.Contains elem @>) |> ignore
         table.PutItem(item, <@ fun r -> r.Set |> Set.contains elem @>) |> ignore
 
+
```
Comment (bartelink) [inline tests/FSharp.AWS.DynamoDB.Tests/ConditionalExpressionTests.fs:335]: remove

File: tests/FSharp.AWS.DynamoDB.Tests/ConditionalExpressionTests.fs [hunk 5]
```diff
@@ -528,4 +531,28 @@ type ``Conditional Expression Tests`` (fixture : TableFixture) =
...
+        let _key = table.PutItem item
+
+        let testScan (expr:Quotations.Expr<('TRecord -> bool)>)=
+            let res = table.Scan expr
+            test <@res.Length=1@>
+
+        testScan <@ fun r -> [|item.Value+10L;item.Value-10L;item.Value|]|>Array.contains r.Value  @>
+        testScan <@ fun r -> elem|>Array.contains r.Value  @>
+        testScan <@ fun r -> elemL|>List.contains r.Value  @>
+    let [<Fact>] ``Table List or Array contains item`` () =
```
Comment (bartelink) [inline tests/FSharp.AWS.DynamoDB.Tests/ConditionalExpressionTests.fs:547]: blank line (and lots of spaces in other places)

File: tests/FSharp.AWS.DynamoDB.Tests/ConditionalExpressionTests.fs [hunk 6]
```diff
@@ -528,4 +531,28 @@ type ``Conditional Expression Tests`` (fixture : TableFixture) =
...
+        let item = mkItem()
+        let _key = table.PutItem item
+
+        let testScan (expr:Quotations.Expr<('TRecord -> bool)>)=
+            let res = table.Scan expr
+            test <@res.Length=1@>
```
Comment (bartelink) [inline tests/FSharp.AWS.DynamoDB.Tests/ConditionalExpressionTests.fs:553]:
```suggestion
            test <@ res.Length = 1 @>
```

**Comment by samritchie** [comment]
Whitespace in general here is a little off - is it worth having a look at putting the standards in the `.editorconfig` and automatically applying (or checking)? Does @fsprojects have any sort of standard formatting?

**Comment by bartelink** [comment]
I forgot to say: nice feature, great that the README and all is included, kudos!

> Does @fsprojects have any sort of standard formatting?

Doubt it. If one comes about, I'll be stealing stuff to put in https://github.com/jet/equinox/blob/master/.editorconfig (maybe that makes sense as a base - it's not exactly extensive, but IME the things I highlighted are the main things that tend to bother me, and doing a cleanup to make the line endings/training whitespace consistent before this PR would be excellent given how much it affects this PR).

File: src/FSharp.AWS.DynamoDB/Expression/ConditionalExpr.fs [hunk 7]
```diff
@@ -192,36 +192,48 @@ let extractQueryExpr (recordInfo : RecordTableInfo) (expr : Expr) : ConditionalE
...
+                |>Seq.toList
```
[the remark this answers is not in the document]
Comment (faldor20) [reply src/FSharp.AWS.DynamoDB/Expression/ConditionalExpr.fs:233]: Well the function returns a ResizeArray and the `In` DU takes a list of Operands:
```
    | In of Operand * Operand list
```
So, we could change this to an `Operand resizeArray` then we never need to do a conversion.
Or we could use an array `Operand array` type and do a conversion.

I did a quick benchmark iterating over some arrays in essentially the same way we are here.
This is with 100 items.
|     Method |     Mean |   Error |   StdDev | Ratio | RatioSD |   Gen0 | Allocated | Alloc Ratio |
|----------- |---------:|--------:|---------:|------:|--------:|-------:|----------:|------------:|
|  ResizeArrayOnly | 339.9 ns | 6.82 ns | 17.25 ns |  1.00 |    0.00 | 0.0725 |     456 B |        1.00 |
| ArrayFirst | 108.1 ns | 2.82 ns |  8.26 ns |  0.32 |    0.03 | 0.1351 |     848 B |        1.86 |

Resizearrays are slower but obviously you don't allocate twice.

However it is worth noting we are unlikely to ever have lists above a pretty small threshold so this doesn't really matter all that much. I'm happy to go with your prefered solution

**Comment by faldor20** [comment]
Thanks! Sorry about the messy formatting, It's just not something I've ever really noticed. 😅

**Comment by faldor20** [comment]
Okay @bartelink I implemented the most performant solution available. I had to replace the nice pattern matching `writeops`, but it's now as fast and memory efficient as it can be.

File: tests/FSharp.AWS.DynamoDB.Tests/ConditionalExpressionTests.fs [hunk 8]
```diff
@@ -531,28 +530,27 @@ type ``Conditional Expression Tests`` (fixture : TableFixture) =
...
-        let elemL = [item.Value+10L;item.Value-10L;item.Value]
+        let elem = [| item.Value+10L; item.Value-10L; item.Value |]
+        let elemL = [ item.Value+10L; item.Value-10L; item.Value ]
+
         let _key = table.PutItem item
 
-        let testScan (expr:Quotations.Expr<('TRecord -> bool)>)=
-            let res = table.Scan expr
-            test <@res.Length=1@>
+        testScan 1 <@ fun r -> [| item.Value+10L; item.Value-10L; item.Value |]|>Array.contains r.Value  @>
```
Comment (bartelink) [inline tests/FSharp.AWS.DynamoDB.Tests/ConditionalExpressionTests.fs:544]:
```suggestion
        testScan 1 <@ fun r -> [| item.Value + 10L; item.Value - 10L; item.Value |] |>Array.contains r.Value @>
```

File: tests/FSharp.AWS.DynamoDB.Tests/ConditionalExpressionTests.fs [hunk 9]
```diff
@@ -531,28 +530,27 @@ type ``Conditional Expression Tests`` (fixture : TableFixture) =
...
+        testScan 1 <@ fun r -> elem|>Array.contains r.Value  @>
+        testScan 1 <@ fun r -> elemL|>List.contains r.Value  @>
```
Comment (bartelink) [inline tests/FSharp.AWS.DynamoDB.Tests/ConditionalExpressionTests.fs:546]:
```suggestion
        testScan 1 <@ fun r -> elemL |> List.contains r.Value @>
```

File: src/FSharp.AWS.DynamoDB/Expression/ConditionalExpr.fs [hunk 10]
```diff
@@ -30,7 +30,7 @@ type QueryExpr =
     | Or of QueryExpr * QueryExpr
     | Compare of Comparator * Operand * Operand
     | Between of Operand * Operand * Operand
-    | In of Operand * Operand list
+    | In of Operand * Operand array
```
Comment (bartelink) [inline src/FSharp.AWS.DynamoDB/Expression/ConditionalExpr.fs:33]: Per F# style quide
```suggestion
    | In of Operand * Operand[]
```

File: src/FSharp.AWS.DynamoDB/Expression/ConditionalExpr.fs [hunk 11]
```diff
@@ -192,36 +192,47 @@ let extractQueryExpr (recordInfo : RecordTableInfo) (expr : Expr) : ConditionalE
...
             | _ -> invalidQuery()
 
+        ///Attempts to convert an expression into a list of operands.
+        ///This should be used when we know the expression is some kind of sequence.
+        let (|SequenceGet|_|) (expr:Expr)=
+            let op = extractOperand None expr
+            match op with
+            |Value va ->
+                va.AttributeValue.L
+                |>ResizeArray.mapToArray (fun x-> x |> wrap |> Value)
```
Comment (bartelink) [inline src/FSharp.AWS.DynamoDB/Expression/ConditionalExpr.fs:232]:
```suggestion
                |> ResizeArray.mapToArray (fun x-> x |> wrap |> Value)
```

**Comment by samritchie** [comment]
Sigh, more fake-related breakages. It looks like there’s a straightforward fsi-related fix, I’ll have a go at this.

**Comment by samritchie** [comment]
@faldor20 I’ve fixed the fake build with #69 if you want to merge that into this PR - thanks

**Comment by faldor20** [comment]
@samritchie @bartelink If one of you could approve this that would be great :)

**Review by samritchie — APPROVED** [review]

**Pull request merged**